src/Raven.Client/Documents/Session/Operations/BatchOperation.cs

```diff
@@ -77,21 +80,154 @@ CommandType GetCommandType(BlittableJsonReaderObject batchResult)
+                        break;
+                    case CommandType.PATCH:
+                        HandlePatch(batchResult);
+                        break;
+                    case CommandType.AttachmentPUT:
+                        HandleAttachmentPut(batchResult);
+                        break;
+                    case CommandType.AttachmentDELETE:
+                        HandleAttachmentDelete(batchResult);
+                        break;
+                    case CommandType.AttachmentMOVE:
+                        HandleAttachmentMove(batchResult);
+                        break;
+                    case CommandType.AttachmentCOPY:
+                        HandleAttachmentCopy(batchResult);
+                        break;
+                    case CommandType.CompareExchangePUT:
+                    case CommandType.CompareExchangeDELETE:
+                        break;
+                    case CommandType.Counters:
+                        HandleCounters(batchResult);
+                        break;
+                    default:
+                        throw new NotSupportedException($"Command '{type}' is not supported.");
                 }
             }
+        }
+
+        private void HandleAttachmentCopy(BlittableJsonReaderObject batchResult)
+        {
+            HandleAttachmentPutInternal(batchResult, CommandType.AttachmentCOPY, nameof(CopyAttachmentCommandData.Id), nameof(CopyAttachmentCommandData.Name));
+        }
+
+        private void HandleAttachmentMove(BlittableJsonReaderObject batchResult)
+        {
+            HandleAttachmentDeleteInternal(batchResult, CommandType.AttachmentMOVE, nameof(MoveAttachmentCommandData.Id), nameof(MoveAttachmentCommandData.Name));
+            HandleAttachmentPutInternal(batchResult, CommandType.AttachmentMOVE, nameof(MoveAttachmentCommandData.DestinationId), nameof(MoveAttachmentCommandData.DestinationName));
+        }
+
+        private void HandleAttachmentDelete(BlittableJsonReaderObject batchResult)
+        {
+            HandleAttachmentDeleteInternal(batchResult, CommandType.AttachmentDELETE, Constants.Documents.Metadata.Id, nameof(DeleteAttachmentCommandData.Name));
+        }
+
+        private void HandleAttachmentDeleteInternal(BlittableJsonReaderObject batchResult, CommandType type, string idFieldName, string attachmentNameFieldName)
+        {
+            var id = GetStringField(batchResult, type, idFieldName);
 
-            // DeferredCommands
-            for (var i = _sessionCommandsCount; i < result.Results.Length; i++)
+            if (_session.DocumentsById.TryGetValue(id, out var documentInfo) == false)
+                return;
+
+            if (documentInfo.Metadata.TryGet(Constants.Documents.Metadata.Attachments, out BlittableJsonReaderArray attachmentsJson) == false || attachmentsJson == null ||
+                attachmentsJson.Length == 0)
+                return;
+
+            var name = GetStringField(batchResult, type, attachmentNameFieldName);
```

Comment:
We could have a direct comparison between this and the name from Attachment in line 150